kernel_test.go

```diff
@@ -0,0 +1,87 @@
+package cbpfc
+
+import (
+	"bytes"
+	"io/ioutil"
+	"net"
+	"os"
+	"testing"
+	"time"
+	"unsafe"
+
+	"golang.org/x/net/bpf"
+	"golang.org/x/sys/unix"
+)
+
+// kernelBackend is a backend that runs cBPF in the kernel
+func kernelBackend(tb testing.TB, insns []bpf.Instruction, in []byte) XDPAction {
+	filter, err := bpf.Assemble(insns)
+	if err != nil {
+		tb.Fatal(err)
+	}
+
+	// Use a unix socket to test the filter
+	// This doesn't risk interfering with any other network traffic, doesn't require / add special
+	// headers (as would be the case if we used UDP for example) that the XDP tests don't deal with,
+	// and doesn't require any special permissions
+
+	// Temporary file for unix socket
+	temp, err := ioutil.TempFile("", "cbpfc_test")
+	if err != nil {
+		tb.Fatal(err)
+	}
+	path := temp.Name()
+	if err := os.Remove(path); err != nil {
+		tb.Fatal(err)
+	}
+
+	read, err := net.ListenUnixgram("unixgram", &net.UnixAddr{Name: path, Net: "unixgram"})
+	if err != nil {
+		tb.Fatal(err)
+	}
+	readConn, err := read.SyscallConn()
+	if err != nil {
+		tb.Fatal(err)
+	}
+	err = readConn.Control(func(fd uintptr) {
+		err := unix.SetsockoptSockFprog(int(fd), unix.SOL_SOCKET, unix.SO_ATTACH_FILTER, &unix.SockFprog{
+			Len:    uint16(len(filter)),
+			Filter: (*unix.SockFilter)(unsafe.Pointer(&filter[0])),
+		})
+		if err != nil {
+			tb.Fatal(err)
+		}
+	})
+	if err != nil {
+		tb.Fatal(err)
+	}
+
+	write, err := net.Dial("unixgram", path)
+	if err != nil {
+		tb.Fatal(err)
+	}
+
+	if _, err := write.Write(in); err != nil {
+		tb.Fatal(err)
+	}
+
+	read.SetDeadline(time.Now().Add(50 * time.Millisecond))
+
+	var out [1500]byte
+	n, err := read.Read(out[:])
+	if err != nil {
+		// A timeout indicates that the filter did not match
```

Comment:
I kind of expected it to be the other way round. If the filter matches, the packet is dropped. As a result we get a timeout?

Reply:
It's unfortunately confusing, the XDP programs we build drop matching packets, but SocketFilters drop packets that do not match.

I reworked this so backends return a `bool` reporting if the packet matched the filter or not instead of `XDPAction` as I think it makes this case less confusing, and added a better comment:

> // SocketFilters only allow matching packets through
// If the packet does not match, the only signal we have is the absence of a packet

Reply:
Personally I find the bool more difficult. Is it `dropped bool` or `accepted bool`?

Reply:
It's neither `dropped` nor `accepted`, that's what I was trying to express with the bool :P
We're interested in knowing if the packet matches the filter or not.
Depending on the backend, "dropped" or "accepted" mean different things:

* In the generated XDP programs for the C and eBPF backends, we return `XDP_DROP` to indicate the packet matched.
* For SocketFilters, the kernel drops packets that **do not** match.

The bool just indicates if the packet matched. Each backend has to interpret whatever signal it gets appropriately.

Reply:
Yeah that makes sense, but it kind of illustrates my point. Thinking about unnamed bools is difficult.

Reply:
Changed it to return a `result` type, with two values `match` and `noMatch`.